### PR TITLE
V8: Embed models builder

### PR DESCRIPTION
**Pull request by bergmania**
#### Changes by Stephane:
- Removed the nuget reference to ModelsBuilder
- Removed dependency to roslyn
- Removed support for custom attributes
- Removed support for modifying models
- Removed support for events
- Removed stuff to support API (`src/Umbraco.ModelsBuilder/Api/**/*`)
- Removed support for dll model (`LiveDll` and `Dll`)

#### Changes by me:
- Updated the `GetModelContentType` and `GetModelPropertyType` from the auto-generated files to return interfaces instead of concrete types.
- Added overload of `GetModelPropertyType` that takes interface instead of concreate type, and obsolete the overload with concrete type.
- Removed unused classes `ClrNameSource` and  `EnumerableExtensions`, such that we should not support these
- Internalized `TextHeaderWriter`
- Fixed build pipeline issues
  - `Microsoft.CodeAnalysis.CSharp` version aligned with the rest of the projects
  - Specified the C# version used in the `Umbraco.ModelsBuilder` project, to `7.3` to support the features used
- Refactored use of `new Exception(...)` to more explicit exception types.
- Removed option to disable ModelsBuilder. Do not make sense for V8 to allow this, and everything is broken if you do that today.

#### Architecture:
- Interfaces to make it possible to replace with a external modelsbuild already exists in `Umbraco.Core`
- Other ModelsBuilder stuff is placed in `Umbraco.ModelsBuilder` that is also added to the `Umbraco.Web` nuget package.

### CODE REVIEW

File: src/Umbraco.ModelsBuilder/Building/TypeModel.cs [hunk 1]
```diff
@@ -204,5 +185,21 @@ public IEnumerable<TypeModel> EnumerateBaseTypes(bool andSelf = false)
                 typeModel = typeModel.BaseType;
             }
         }
+
+        /// <summary>
+        /// Maps ModelType.
+        /// </summary>
+        public static void MapModelTypes(IList<TypeModel> typeModels, string ns)
```
Comment (Shazwazza) [inline src/Umbraco.ModelsBuilder/Building/TypeModel.cs]: leave this as public or internal?

File: src/Umbraco.ModelsBuilder/Dashboard/BuilderDashboardHelper.cs [hunk 2]
```diff
@@ -1,21 +1,17 @@
...
 
 namespace Umbraco.ModelsBuilder.Dashboard
 {
     internal static class BuilderDashboardHelper
     {
-        public static bool CanGenerate()
-        {
-            return UmbracoConfig.For.ModelsBuilder().ModelsMode.SupportsExplicitGeneration();
-        }
+        private static Config Config => Current.Configs.ModelsBuilder();
```
Comment (Shazwazza) [inline src/Umbraco.ModelsBuilder/Dashboard/BuilderDashboardHelper.cs]: inject? means making this not static, or passing this in to the methods, also ideally we'd find a better name for this than "Helper"

File: src/Umbraco.ModelsBuilder/ReferencedAssemblies.cs [hunk 3]
```diff
@@ -31,19 +25,68 @@ static ReferencedAssemblies()
...
+
+            return null;
+        }
+
+        public static Assembly GetNetStandardAssembly()
+        {
+            // in PreApplicationStartMethod we cannot get BuildManager.Referenced assemblies, do it differently
+            try
+            {
+                var someAssembly = Assembly.Load("System.Reflection.Metadata");
```
Comment (Shazwazza) [inline src/Umbraco.ModelsBuilder/ReferencedAssemblies.cs]: code duplication, should prob be shared properly

File: src/Umbraco.ModelsBuilder/Umbraco/HashHelper.cs [hunk 4]
```diff
@@ -6,13 +6,10 @@ namespace Umbraco.ModelsBuilder.Umbraco
 {
     class HashHelper
```
Comment (Shazwazza) [inline src/Umbraco.ModelsBuilder/Umbraco/HashHelper.cs]: "Helper", could do with a better name

File: src/Umbraco.ModelsBuilder/Umbraco/PureLiveModelFactory.cs [hunk 5]
```diff
@@ -300,19 +328,12 @@ internal Infos EnsureModels()
 
         private Assembly GetModelsAssembly(bool forceRebuild)
         {
-            var modelsDirectory = UmbracoConfig.For.ModelsBuilder().ModelsDirectory;
+            var modelsDirectory = _config.ModelsDirectory;
             if (!Directory.Exists(modelsDirectory))
                 Directory.CreateDirectory(modelsDirectory);
 
-            // must filter out *.generated.cs because we haven't deleted them yet!
-            var ourFiles = Directory.Exists(modelsDirectory)
```
Comment (Shazwazza) [inline src/Umbraco.ModelsBuilder/Umbraco/PureLiveModelFactory.cs]: I wonder why we are not using this as part of the hash any longer? Seems like an odd change since this is most likely how it is working in the original MB

File: src/Umbraco.ModelsBuilder/Umbraco/LiveModelsProvider.cs [hunk 6]
```diff
@@ -21,15 +21,10 @@ public sealed class LiveModelsProvider
...
-        internal static bool IsEnabled
-        {
-            get
-            {
-                var config = UmbracoConfig.For.ModelsBuilder();
-                return config.ModelsMode.IsLiveNotPure();
-                // we do not manage pure live here
-            }
-        }
+        private static Config Config => Current.Configs.ModelsBuilder();
```
Comment (Shazwazza) [inline src/Umbraco.ModelsBuilder/Umbraco/LiveModelsProvider.cs]: inject? means less statics

File: src/Umbraco.ModelsBuilder/Umbraco/LiveModelsProvider.cs [hunk 7]
```diff
@@ -133,9 +127,7 @@ public void Dispose()
 
         public static void Install()
         {
-            if (!LiveModelsProvider.IsEnabled)
-                return;
-
+            // always - don't read config in PreApplicationStartMethod
             HttpApplication.RegisterModule(typeof(LiveModelsProviderModule));
```
Comment (Shazwazza) [inline src/Umbraco.ModelsBuilder/Umbraco/LiveModelsProvider.cs]: This is what we'll most likely have to alter a bit to detect the original MB is active.

File: src/Umbraco.ModelsBuilder/Umbraco/ModelsBuilderInitializer.cs [hunk 8]
```diff
@@ -0,0 +1,28 @@
+﻿using System.Web;
+using System.Web.Compilation;
+using Umbraco.ModelsBuilder.Umbraco;
+
+[assembly: PreApplicationStartMethod(typeof(ModelsBuilderInitializer), "Initialize")]
+
+namespace Umbraco.ModelsBuilder.Umbraco
+{
+    public static class ModelsBuilderInitializer
```
Comment (Shazwazza) [inline src/Umbraco.ModelsBuilder/Umbraco/ModelsBuilderInitializer.cs]: doesn't need to be public

File: src/Umbraco.ModelsBuilder/Umbraco/ModelsGenerationError.cs [hunk 9]
```diff
@@ -1,13 +1,15 @@
...
 using System.Text;
-using Umbraco.Core.Configuration;
+using Umbraco.Core.Composing;
 using Umbraco.ModelsBuilder.Configuration;
 
 namespace Umbraco.ModelsBuilder.Umbraco
 {
     internal static class ModelsGenerationError
     {
+        private static Config Config => Current.Configs.ModelsBuilder();
```
Comment (Shazwazza) [inline src/Umbraco.ModelsBuilder/Umbraco/ModelsGenerationError.cs]: inject?

File: src/Umbraco.ModelsBuilder/Umbraco/OutOfDateModelsStatus.cs [hunk 10]
```diff
@@ -1,19 +1,18 @@
...
+﻿using System.IO;
+using Umbraco.Core.Composing;
 using Umbraco.ModelsBuilder.Configuration;
 using Umbraco.Web.Cache;
 
 namespace Umbraco.ModelsBuilder.Umbraco
 {
     public sealed class OutOfDateModelsStatus
     {
+        private static Config Config => Current.Configs.ModelsBuilder();
```
Comment (Shazwazza) [inline src/Umbraco.ModelsBuilder/Umbraco/OutOfDateModelsStatus.cs]: inject?

File: src/Umbraco.ModelsBuilder/Umbraco/PublishedModelUtility.cs [hunk 11]
```diff
@@ -40,7 +41,15 @@ public static PublishedContentType GetModelContentType(PublishedItemType itemTyp
             }
         }
 
-        public static PublishedPropertyType GetModelPropertyType<TModel, TValue>(PublishedContentType contentType, Expression<Func<TModel, TValue>> selector)
+        [Obsolete("Use the overload taking IPublishedContentType instead")]
```
Comment (Shazwazza) [inline src/Umbraco.ModelsBuilder/Umbraco/PublishedModelUtility.cs]: Do we need to obsolete anything here? This is brand new code/APIs that has never been shipped so we shouldn't really need to ship an obsoleted method.

File: src/Umbraco.ModelsBuilder/Umbraco/PureLiveModelFactory.cs [hunk 12]
```diff
@@ -24,14 +23,14 @@
 
 namespace Umbraco.ModelsBuilder.Umbraco
```
Comment (Shazwazza) [inline src/Umbraco.ModelsBuilder/Umbraco/PureLiveModelFactory.cs]: Just noticed this namespace, it's not very pretty and IMO we should change this.

File: src/Umbraco.ModelsBuilder/Umbraco/PureLiveModelFactory.cs [hunk 13]
```diff
@@ -182,14 +197,27 @@ private void RazorBuildProvider_CodeGenerationStarted(object sender, EventArgs e
...
+
+                var modelsDirectory = _config.ModelsDirectory;
+                if (!Directory.Exists(modelsDirectory))
+                    Directory.CreateDirectory(modelsDirectory);
+
+                // clear stuff
+                var modelsHashFile = Path.Combine(modelsDirectory, "models.hash");
+                //var modelsSrcFile = Path.Combine(modelsDirectory, "models.generated.cs");
+                //var projFile = Path.Combine(modelsDirectory, "all.generated.cs");
+                var dllPathFile = Path.Combine(modelsDirectory, "all.dll.path");
```
Comment (Shazwazza) [inline src/Umbraco.ModelsBuilder/Umbraco/PureLiveModelFactory.cs]: This seems odd - not sure if this is intended but since we are supporting PureLive and LiveAppData, there is no "DLL" but there is generated code, need to investigate to see if this is a mistake. Same goes with removing the files from being hashed.

File: src/Umbraco.ModelsBuilder/Umbraco.ModelsBuilder.csproj [hunk 14]
```diff
@@ -110,14 +92,23 @@
     </PackageReference>
   </ItemGroup>
   <ItemGroup>
+    <ProjectReference Include="..\Umbraco.Abstractions\Umbraco.Abstractions.csproj">
```
Comment (Shazwazza) [inline src/Umbraco.ModelsBuilder/Umbraco.ModelsBuilder.csproj]: Hrm, this isn't a project ;) I'll fixup

File: src/Umbraco.ModelsBuilder/Umbraco/LiveModelsProvider.cs [hunk 15]
```diff
@@ -133,9 +127,7 @@ public void Dispose()
 
         public static void Install()
         {
-            if (!LiveModelsProvider.IsEnabled)
```
Comment (Shazwazza) [inline src/Umbraco.ModelsBuilder/Umbraco/LiveModelsProvider.cs]: Seems odd this was removed, i'll update with recent changes. The module should not be enabled if pure live is not enabled.

File: src/Umbraco.ModelsBuilder/Umbraco/PublishedModelUtility.cs (hunk 11 above)
Comment (bergmania) [reply src/Umbraco.ModelsBuilder/Umbraco/PublishedModelUtility.cs]: You're right.. We don't..

File: src/Umbraco.ModelsBuilder/Umbraco/PureLiveModelFactory.cs (hunk 12 above)
Comment (bergmania) [reply src/Umbraco.ModelsBuilder/Umbraco/PureLiveModelFactory.cs]: Do you have any suggestions?

File: src/Umbraco.ModelsBuilder/Umbraco.ModelsBuilder.csproj (hunk 14 above)
Comment (bergmania) [reply src/Umbraco.ModelsBuilder/Umbraco.ModelsBuilder.csproj]: whoops.. (I tested on the unicore project, must have forgot to revert that change)

File: src/Umbraco.ModelsBuilder/Umbraco/PureLiveModelFactory.cs (hunk 5 above)
Comment (Shazwazza) [reply src/Umbraco.ModelsBuilder/Umbraco/PureLiveModelFactory.cs]: I don't believe this should have been removed, I will re-implement these checks in my new branch.

Comment (Shazwazza) [reply src/Umbraco.ModelsBuilder/Umbraco/PureLiveModelFactory.cs]: Actually, on second thought, the reason this is removed is because it's not really including any files since the only files that will exist in this folder that do not end with .generated.cs is just the hash file. So i guess inluding these files is useless. Perhaps this is for when DLL mode is enabled, but it seems not relevant for either purelive appdata or liveappdata.

File: src/Umbraco.ModelsBuilder/Building/Builder.cs [hunk 16]
```diff
@@ -273,20 +201,10 @@ internal void PrepareAmbiguousSymbols()
...
-
-            // only 1 - ensure it matches
-            var found = symbols[0].ToDisplayString();
-            var pos = found.IndexOf('<'); // generic?
-            if (pos > 0) found = found.Substring(0, pos); // strip
-            return found != match; // and compare
+            // cannot figure out is a symbol is ambiguous without Roslyn
+            // so... let's say everything is ambiguous - code won't be
+            // pretty but it'll work
+            return true;
```
[the remark this answers is not in the document]
Comment (Shazwazza) [reply src/Umbraco.ModelsBuilder/Building/Builder.cs]: I've added notes to this in the code, it just means that the generated csharp will have `global::` syntax for most of it's types.

File: src/Umbraco.ModelsBuilder/Umbraco/PureLiveModelFactory.cs (hunk 13 above)
Comment (Shazwazza) [reply src/Umbraco.ModelsBuilder/Umbraco/PureLiveModelFactory.cs]: This is how the source is in the orig MB too so it's been working like that for a long time. The reason is because we don't need to take into account these commented out files when checking if models need to be rebuilt and they can be left there for reference just in case.

File: src/Umbraco.Web/Properties/AssemblyInfo.cs [hunk 17]
```diff
@@ -12,7 +12,7 @@
 
 // Umbraco Cms
 [assembly: InternalsVisibleTo("Umbraco.Web.UI")]
-
+[assembly: InternalsVisibleTo("Umbraco.ModelsBuilder.Embedded")]
```
Comment (zpqrtbnk) [inline src/Umbraco.Web/Properties/AssemblyInfo.cs]: why is this needed?

File: src/Umbraco.Core/Properties/AssemblyInfo.cs [hunk 18]
```diff
@@ -14,6 +14,7 @@
 [assembly: InternalsVisibleTo("Umbraco.Web")]
 [assembly: InternalsVisibleTo("Umbraco.Web.UI")]
 [assembly: InternalsVisibleTo("Umbraco.Examine")]
+[assembly: InternalsVisibleTo("Umbraco.ModelsBuilder.Embedded")]
```
Comment (zpqrtbnk) [inline src/Umbraco.Core/Properties/AssemblyInfo.cs]: why is this needed?

Comment (bergmania) [reply src/Umbraco.Core/Properties/AssemblyInfo.cs]: We use `PanicException` + `UrlExtensions.IsClientSideRequest` from Core

File: src/Umbraco.Web/Properties/AssemblyInfo.cs (hunk 17 above)
Comment (bergmania) [reply src/Umbraco.Web/Properties/AssemblyInfo.cs]: We use `UmbracoFeatures.Disabled` to removed the embedded modelsbuilder dashboard if we detect the full version.

File: src/Umbraco.Core/Properties/AssemblyInfo.cs (hunk 18 above)
Comment (zpqrtbnk) [reply src/Umbraco.Core/Properties/AssemblyInfo.cs]: ok

File: src/Umbraco.Web/Properties/AssemblyInfo.cs (hunk 17 above)
Comment (zpqrtbnk) [reply src/Umbraco.Web/Properties/AssemblyInfo.cs]: never mind, found it. this should remove the dashboard and the controller:

    // remove the embedded dashboard
    composition.Dashboards().Remove<ModelsBuilderDashboard>();

    // remove the embedded controller
    composition.WithCollectionBuilder<UmbracoApiControllerTypeCollectionBuilder>()
        .Remove<ModelsBuilderDashboardController>();

**Comment by warrenbuckley** [comment]
## Tested
* Vanilla 8.5 Install with SK & various MB modes
* Tested an 8.4 to 8.5 upgrade (with new & old DLLs)
* Tested by removing old DLL after upgrade
* Tested doing fundamentals course on 8.5

**Comment by dawoe** [comment]
@bergmania 

I know I'm late to the party but....

Where was the RFC for this ? So we as community members who use MB on a daily basis could have given input on how they use it.

With this PR modelsbuilder has become very limited in functionality. 

What was the reasoning behind this ?

Dave

**Comment by bergmania** [comment]
Hi @dawoe

You are able to use the existing (external) models builder with full functionality. You just need to add the nuget package. We detect the dll.

**Comment by Box-Of-Hats** [comment]
Are there any blogs posts explaining more about the shift for the ModelBuilder? I've had a look around and I haven't been able to find any reasoning behind it or how this will affect me as an Umbraco developer

**Comment by nul800sebastiaan** [comment]
Please note that the full version of ModelsBuilder isn't going anywhere, you can keep using that in exactly the same way. 

We talked about this in a bit more detail in the latest product update (scroll to "Umbraco 8.5"): https://umbraco.com/blog/umbraco-product-update-december-18th-2019/

**Comment by lukehook** [comment]
It's mentioned that installing the full version of models builder is still supported and that the .dll is detected. I've just installed the models builder from nuget and now I'm getting an error that Dll mode is on a valid option. What else do I need to do to get the previous models builder to work or is this not possible afterall?

**Comment by bergmania** [comment]
Hi @lukehook 

Can you confirm the `Umbraco.ModelsBuilder.dll` is in your bin folder?

When I try to reproduce, I can't get any error when the `Umbraco.ModelsBuilder.dll`  is in the bin folder. But if I delete this dll - and thereby running on the embedded, I get the following error:
```
System.Configuration.ConfigurationErrorsException: ModelsMode "Dll" is not a valid mode. Note that modes are case-sensitive. Possible values are: Nothing, PureLive, AppData, LiveAppData
```

**Comment by lukehook** [comment]
@bergmania sorry I should have been more specific on when I'm seeing the error. The dll is definitely in the /bin folder. I'm seeing the error whenever I attempt to save a doc type. I can generate the models absolutely find in the ModelsBuilder dashboard. But when I save a doc type I'm seeing the error you are referring to

**Comment by bergmania** [comment]
Thanks @lukehook, now I was able to reproduce 😢
